ci-tests.sh

```diff
@@ -0,0 +1,116 @@
+#!/bin/bash
+set -euo pipefail
+# In CI, we do not have TLS/DNS set up. Docker requires TLS, so we cannot do
+# proper functional tests until post deployment. And most tools require DNS in
+# order to hit the right vhost in nginx.
+#
+# So we use curl's --resolve opti0on to query the base APIs to check we get the
+# appropriate responses at the http level.
+#
+# see also: full-tests.sh, which can be run post deploy 
+# 
+url=
+body=$(mktemp)
+headers=$(mktemp)
+return_code=0
+
+last_test_failed=0
+
+fail() {
+    echo "FAIL: $*"
+    last_test_failed=1
+    return_code=1
+    cat "$headers"
+    cat "$body"
+}
+
+ok() {
+    echo "OK: $*"
+    last_test_failed=0
+}
+
+try() {
+    # reset tests
+    last_test_failed=0
+    # global
+    url=$1
+    local expected=$2
+
+    code="$(
+        curl --verbose --output "$body" \
+        --resolve github-proxy.opensafely.org:80:127.0.0.1 \
+        --resolve docker-proxy.opensafely.org:80:127.0.0.1 \
+        --write-out "%{http_code}"\
+        "$url" \
+        2> "$headers"
+    )"
+
+    if test "$code" != "$expected"; then
+        fail "$url returned $code, not $expected"
+    else
+        ok "$url returned $code"
+    fi
+}
+
+
+assert-body() {
+    if test "$last_test_failed" = "1"; then
+        echo "SKIP assert body"
+        return
+    fi
+    local expected=$1
+    if test "$(cat "$body")" == "$expected"; then
+        ok "$url returned expected body"
+    else
+        fail "$url returned unexpected body:"
```

Comment:
should this have a `$body` at the end to output the bad body?

Reply:
the fail() function outputs the body by default?

Reply:
ah, the trailing :, will fix